libia2/ia2.c

```diff
@@ -123,49 +143,44 @@ int protect_pages(struct dl_phdr_info *info, size_t size, void *data) {
 
   for (size_t i = 0; i < info->dlpi_phnum; i++) {
     Elf64_Phdr phdr = info->dlpi_phdr[i];
-    if ((phdr.p_type != PT_LOAD) && (phdr.p_type != PT_GNU_RELRO)) {
+    if ((phdr.p_type != PT_LOAD)) {
+      continue;
+    }
+
+    if ((phdr.p_flags & PF_W) == 0) {
+      // We don't need to protect any read-only memory that was mapped in from
+      // the object file. This data is not assumed to be secret and can be
+      // shared by default.
       continue;
     }
 
-    int prot = segment_flags_to_access_flags(phdr.p_flags);
+    int access_flags = segment_flags_to_access_flags(phdr.p_flags);
 
     Elf64_Addr start = (info->dlpi_addr + phdr.p_vaddr) & ~0xFFFUL;
-    Elf64_Addr stop = (start + phdr.p_memsz + 0xFFFUL) & ~0xFFFUL;
-    // TODO: This logic assumes that each segment may only be split by one
-    // shared section
-    for (size_t j = 0; j < NUM_SHARED_RANGES; j++) {
-      if ((shared_ranges[j].start <= start) &&
-          (shared_ranges[j].end >= start)) {
-        start = shared_ranges[j].end;
-      } else if ((shared_ranges[j].end >= stop) &&
-                 (shared_ranges[j].start <= stop)) {
-        stop = shared_ranges[j].start;
-      } else if ((shared_ranges[j].end <= stop) &&
-                 (shared_ranges[j].start >= start)) {
-        size_t len = stop - shared_ranges[j].end;
-        segment_info[extra_seg_info].addr = shared_ranges[j].end;
-        segment_info[extra_seg_info].size = len;
-        segment_info[extra_seg_info].prot = prot;
-
-        stop = shared_ranges[j].start;
-        extra_seg_info += 1;
+    Elf64_Addr seg_end = (start + phdr.p_memsz + 0xFFFUL) & ~0xFFFUL;
+    while (start < seg_end) {
+      Elf64_Addr cur_end = seg_end;
+
+      // Adjust start and cur_end to bound the lowest region between start and
+      // seg_end that is not covered by a shared range
+      for (size_t j = 0; j < shared_range_count; j++) {
+        if (shared_ranges[j].start <= start && shared_ranges[j].end > start) {
+          start = shared_ranges[j].end;
+        } else if (shared_ranges[j].start >= start &&
```

Comment:
Suppose `shared_ranges[j].start` is equal to `start` (to see what happens in the `=` case of this `>=`).

Then, either the condition on L167 holds (and we don't reach the `else if`), or `shared_ranges[j].end` is not greater than `start` (i.e., this is a zero-length range). The body of this `if` is a no-op in the zero-length range case, and it's clearer to not match it in the condition (which is looking for an overlapping, rather than touching, range to skip) than match it and perform a no-op. So I would advocate making this `>=` simply a `>`.

Reply:
Comment this condition: ``// Look for a shared range that starts within [`start`, `cur_end`)``

Reply:
Which >= are you referring to here? I'm making `shared_ranges[j].start < cur_end` in second half of the second conditional.

Reply:
The `else` is unnecessary, I just put it for clarity because if the first conditional triggers the second cannot.

Reply:
Ahhh I see what you mean. Hopefully I've clarified sufficiently now?

Reply:
Yes, looks clear and correct now. :+1: